model/context.go

```diff
@@ -18,107 +18,264 @@
-		if p, err := strconv.Atoi(portStr); err == nil {
-			portInt = &p
-		}
-	}
-	utility.Add(destination, "port", portInt)
-	utility.Add(destination, "original", source("request.url.raw"))
-	if scheme, ok := source("request.url.protocol").(string); ok {
-		utility.Add(destination, "scheme", strings.TrimSuffix(scheme, ":"))
-	}
-	utility.Add(destination, "query", source("request.url.search"))
+type Http struct {
+	Version  *string
+	Request  *Req
+	Response *Resp
+}
 
-	return destination
+type Url struct {
+	Original *string
+	Scheme   *string
+	Full     *string
+	Domain   *string
+	Port     *int
+	Path     *string
+	Query    *string
+	Fragment *string
 }
 
 type Page struct {
 	Url     *string
 	Referer *string
 }
 
-func DecodePage(input interface{}, err error) (*Page, error) {
+type Label common.MapStr
+type Custom common.MapStr
+type Headers common.MapStr
+
+type Req struct {
+	Method  string
+	Body    interface{}
+	Headers *Headers
+	Env     interface{}
+	Socket  *Socket
+	Cookies interface{}
+}
+
+type Socket struct {
+	RemoteAddress *string
+	Encrypted     *bool
+}
+
+type Resp struct {
+	Finished    *bool
+	StatusCode  *int
+	HeadersSent *bool
+	Headers     *Headers
+}
+
+func DecodeContext(input interface{}, err error) (*Context, error) {
 	if input == nil || err != nil {
 		return nil, err
 	}
-	raw, ok := input.(common.MapStr)
+	raw, ok := input.(map[string]interface{})
 	if !ok {
-		return nil, errors.New("Invalid type for fetching Page")
+		return nil, errors.New("Invalid type for fetching Context fields")
+	}
+
+	decoder := utility.ManualDecoder{}
+	ctxInp := decoder.MapStr(raw, "context")
+	userInp := decoder.Interface(ctxInp, "user")
+	err = decoder.Err
+	http, err := decodeHttp(ctxInp, err)
+	url, err := decodeUrl(ctxInp, err)
+	label, err := decodeLabel(ctxInp, err)
+	custom, err := decodeCustom(ctxInp, err)
+	page, err := decodePage(ctxInp, err)
+	user, err := metadata.DecodeUser(userInp, err)
+	if err != nil {
+		return nil, err
+	}
+	return &Context{
+		Http:   http,
+		Url:    url,
+		Label:  label,
+		Page:   page,
+		Custom: custom,
+		User:   user,
+	}, nil
+
+}
+
+func decodeUrl(raw common.MapStr, err error) (*Url, error) {
+	if err != nil {
+		return nil, err
+	}
+
+	decoder := utility.ManualDecoder{}
+	req := decoder.MapStr(raw, "request")
+	if req == nil {
+		return nil, decoder.Err
+	}
+
+	inpUrl := decoder.MapStr(req, "url")
+	url := Url{
+		Original: decoder.StringPtr(inpUrl, "raw"),
+		Full:     decoder.StringPtr(inpUrl, "full"),
+		Domain:   decoder.StringPtr(inpUrl, "hostname"),
+		Path:     decoder.StringPtr(inpUrl, "pathname"),
+		Query:    decoder.StringPtr(inpUrl, "search"),
+		Fragment: decoder.StringPtr(inpUrl, "hash"),
+	}
+	if scheme := decoder.StringPtr(inpUrl, "protocol"); scheme != nil {
+		trimmed := strings.TrimSuffix(*scheme, ":")
+		url.Scheme = &trimmed
+	}
+	if decoder.Err != nil {
+		return nil, decoder.Err
+	}
+
+	if url.Port = decoder.IntPtr(inpUrl, "port"); url.Port != nil {
```

Comment:
I'd prefer an `else if` here to group these two operations on `url.Port` together for future maintenance.